### DISCUSSION TITLE
from max98357 audio data to a2dp_source bluetooth headset

### DISCUSSION THREAD

**Discussion by fmkkaya** [pschatzmann/ESP32-A2DP#631] [Q&A] [answered]
### Problem Description

Hello,
I use the audio library with max98357 and transmit the sound to the speaker with i2s. Is it possible to transmit the data in i2s to bluetooth with the BluetoothA2DPSource library. when I examined your library, I could not see it. does your library support this? would you consider adding it?

### Device Description

Esp32-wroom arduino

### Sketch

```cpp
a2dp_source.start("Galaxy Buds+ (D504)", get_sound_data_from_i2s);
```

### Provide your Version of the EP32 Arduino Core (or the IDF Version)

2.0.14

### I have checked existing issues, discussions and online documentation

- [X] I confirm I have checked existing issues, discussions and online documentation

**Comment by pschatzmann** [comment, marked as answer]
There are plenty of [examples in the AudioTools library](https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/examples-communication/a2dp)

**Comment by fmkkaya** [comment]
Thank you for your answer.
https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/examples-communication/a2dp I have reviewed all of this address but I could not find a suitable library.
I kindly request your support.

**Comment by pschatzmann** [comment]
library ? I think you were looking for a sketch from I2S to A2DP. So that would be

- basic-i2s-a2dp
- stream-i2s-a2dp

You forgot to describe your I2S source device from which you want to read the audio from.

**Comment by fmkkaya** [comment]
Thanks for your info.

I tryed stream-i2s-a2dp and  basic-i2s-a2dp but i cant solved. My example code is  below.it's playing a tone every one second. how can i export music at the same time to bluetooth headset. 


#include <Arduino.h>
#include <driver/i2s.h>
#include <math.h>

#define I2S_NUM I2S_NUM_0
#define SAMPLE_RATE 44100
#define I2S_BCK_IO 26  
#define I2S_WS_IO 27  
#define I2S_DATA_IO 25 

const int frequency = 440;  
const int amplitude = 2000; 
const int duration = 2000;

void playTone(int freq, int duration_ms)
{
  const int samples_per_wave = SAMPLE_RATE / freq;
  int16_t *wave = (int16_t *)malloc(samples_per_wave * sizeof(int16_t));

  for (int i = 0; i < samples_per_wave; i++)
  {
    wave[i] = amplitude * sin(2.0 * M_PI * i / samples_per_wave);
  }

  int total_samples = (SAMPLE_RATE * duration_ms) / 1000;
  for (int i = 0; i < total_samples; i += samples_per_wave)
  {
    size_t bytes_written;
    i2s_write(I2S_NUM, wave, samples_per_wave * sizeof(int16_t), &bytes_written, portMAX_DELAY);
  }

  free(wave);
}

void setup()
{
  Serial.begin(115200);
  i2s_config_t i2s_config = {
      .mode = (i2s_mode_t)(I2S_MODE_MASTER | I2S_MODE_TX),
      .sample_rate = SAMPLE_RATE,
      .bits_per_sample = I2S_BITS_PER_SAMPLE_16BIT,
      .channel_format = I2S_CHANNEL_FMT_ONLY_RIGHT,
      .communication_format = I2S_COMM_FORMAT_I2S_MSB,
      .intr_alloc_flags = 0,
      .dma_buf_count = 8,
      .dma_buf_len = 64,
      .use_apll = false,
      .tx_desc_auto_clear = true,
      .fixed_mclk = 0};

  i2s_pin_config_t pin_config = {
      .bck_io_num = I2S_BCK_IO,
      .ws_io_num = I2S_WS_IO,
      .data_out_num = I2S_DATA_IO,
      .data_in_num = I2S_PIN_NO_CHANGE};

  // I2S başlatma
  i2s_driver_install(I2S_NUM, &i2s_config, 0, NULL);
  i2s_set_pin(I2S_NUM, &pin_config);
  i2s_zero_dma_buffer(I2S_NUM);
}

void loop()
{
  playTone(frequency, duration);
  delay(1000);
}

**Comment by pschatzmann** [comment]
You could write to the A2DPStream (from the AudioTools) at the same time when you call  i2s_write or you write it into a queue and provide the audio data from there to the [A2DP callback](https://github.com/pschatzmann/ESP32-A2DP?tab=readme-ov-file#sending-data-from-a-a2ds-data-source-with-a-callback).

Using the AudioTools you could write to a MultiOutput which sends the audio both to a I2SStream and A2DPStream.

ps. why are you still using 2.0.14 ? The actual release is 3.0.7 with 3.1 being released soon!
If you mess around with this obsolete I2S API, you will never be able to upgrade w/o a major rewrite, so I suggest to use the AudioTools that will work with both 2.x and 3.x.

Please also note that there is some considerable lag in A2DP, so you can write it at the same time, but you will not hear it at the same time.

Your example would correspond to
- streams-generator-a2dp
- basic-generator-a2dp

**Comment by fmkkaya** [comment]
Hi Phil,
Thanks for your information.
I am looking about 4 days for this issue. I have a simple code below that plays the mp3.mp3 file on the sd card every 10 seconds. I have max98357 connected to esp32 and I can get the sound from there. If possible can you help me how I can integrate a2dp into this code to transmit the sound to the bluetooth headset.



#include "Arduino.h"
unsigned long lastTime = 0;
#include "Audio.h"
const int PIN_AUDIO_DOUT = 25;
const int PIN_AUDIO_BCLK = 26;
const int PIN_AUDIO_LRC = 27;
Audio audio;

void setup()
{
  Serial.begin(115200);
  SD.begin(5);
  audio.setPinout(PIN_AUDIO_BCLK, PIN_AUDIO_LRC, PIN_AUDIO_DOUT);

  audio.setVolume(5); 

}

void loop()
{
  audio.loop();
  if (millis() > lastTime)
  {

    lastTime = millis() + 10000;
    audio.connecttoFS(SD, "/contents/mp3.mp3"); 
    Serial.println("loop");
  }
}

**Comment by fmkkaya** [comment]
Audio library is https://github.com/schreibfaul1/ESP32-audioI2S.git

**Reply by pschatzmann** [reply]
I am afraid that you have to lean to use the AudioTools library or ask schreibfaul1 for help !
